### ISSUE TITLE
Wait until angular ready

### ISSUE THREAD

**Issue by Alexml8456** [codeborne/selenide#525]
Hi,
Please help me with such difficult issue, regarding to "Angular" application.
In fact "Selenide" almost resolve all asynchronous javascript issues.
But in some, very special, cases I have difficulties with working in "Angular" application. 
So, I found some custom library that completely resolve all javascript issues in Angular.
But this library written in python. And implemented in [Robot Framework](http://robotframework.org/) in [ExtendedSelenium2Library](https://github.com/rickypc/robotframework-extendedselenium2library).
Function that resolve javascript issue given bellow:

```
def wait_until_angular_ready(self, timeout=None, error=None):
    """Waits until [https://goo.gl/Kzz8Y3|AngularJS] is ready to process the next request or
    ``timeout`` expires.

    Arguments:
    - ``timeout``: The maximum value to wait for [https://goo.gl/Kzz8Y3|AngularJS]
                   to be ready to process the next request.
                   See `introduction` for more information about ``timeout`` and
                   its default value.
    - ``error``: The value that would be use to override the default error message.

    Examples:
    | Wait Until Angular Ready | 15s |
    """
    # pylint: disable=no-member
    timeout = self._get_timeout_value(timeout, self._implicit_wait_in_secs)
    if not error:
        error = 'AngularJS is not ready in %s' % self._format_timeout(timeout)
    # we add more validation here to support transition
    # between AngularJs to non AngularJS page.
    # pylint: disable=no-member
    script = self.NG_WRAPPER % {'prefix': 'var cb=arguments[arguments.length-1];'
                                          'if(window.angular){',
                                'handler': 'function(){cb(true)}',
                                'suffix': '}else{cb(true)}'}
    # pylint: disable=no-member
    browser = self._current_browser()
    browser.set_script_timeout(timeout)
    # pylint: disable=bare-except
    try:
        WebDriverWait(browser, timeout, self._inputs['poll_frequency']). \
            until(lambda driver: driver.execute_async_script(script), error)
    except TimeoutException:
        # prevent double wait
        pass
    except:
        self._debug(exc_info()[0])
        # still inflight, second chance. let the browser take a deep breath...
        sleep(self._inputs['browser_breath_delay'])
        try:
            WebDriverWait(browser, timeout, self._inputs['poll_frequency']). \
                until(lambda driver: driver.execute_async_script(script), error)
        except:
            # instead of halting the process because AngularJS is not ready
            # in <TIMEOUT>, we try our luck...
            self._debug(exc_info()[0])
        finally:
            browser.set_script_timeout(self._timeout_in_secs)
    finally:
        browser.set_script_timeout(self._timeout_in_secs)
```
Maybe you have some ideas in what way I have chance to implement such method in Selenide?
Thanks in advance for response.

**Comment by rmarinsky** [comment]
You can use any of ways from this post http://stackoverflow.com/questions/25062969/testing-angularjs-with-selenium

**Comment by Alexml8456** [comment]
Thanks for response! 
I created such method:


    public static void waitForAngularLoad() {
        final String angularReadyScript = "return (window.angular !== undefined) && (angular.element(document.body).injector() !== undefined) && (angular.element(document.body).injector().get('$http').pendingRequests.length === 0)";
        ExpectedCondition<Boolean> angularLoad = webDriver -> Boolean.valueOf(executeJavaScript(angularReadyScript).toString());
        Wait().until(angularLoad);
    }

and it's works perfectly.

**Comment by Alexml8456** [comment]
But, in such implementation, I'll need to call this method before each click (Button or Element) or before Select checkbox or something from list.
May be somebody know how to call this method before each actions automatically?

**Comment by rmarinsky** [comment]
Create some `class CustomEventListener implements WebDriverEventListener`
and in tale `beforeFindBy(...)` put your code
then use `WebDriverRunner.addListener(new CustomEventListener)`

**Comment by Alexml8456** [comment]
Thank alot for help.
Now it works as it should be.

**Comment by silramos** [comment]
It worked here too, but, instead of `"return (window.angular !== undefined) && (angular.element(document.body).injector() !== undefined) && (angular.element(document.body).injector().get('$http').pendingRequests.length === 0)"`, i used the Angular's testability API. So i used `"return window.getAllAngularTestabilities().findIndex(x=>!x.isStable()) === -1"`. 
To see more about this API: https://angular.io/api/core/Testability

Thank you a lot, i was stucked on that.

**Comment by mac2000** [comment]
There are issues with `isStable` it depends on many things which even might be not under control of front end engineers (e.g. timers in 3rd party libraries will brake this)

There is another way - to check pending requests and tasks, in our case we ended up with something like:

```java
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.support.ui.ExpectedCondition;

/**
 * https://github.com/selenide/selenide/issues/525
 */
public class AngularConditions {
    /**
     * Will wait till there is no pending requests or tasks
     * IMPORTANT: It wont work with non angular pages!
     *
     * @return Boolean true when angular is ready, false otherwise
     */
    public static ExpectedCondition<Boolean> angularHasFinishedProcessing() {
        return driver -> {
            var javaScriptExecutor = (JavascriptExecutor) driver;
            String javaScriptExecutionResult = javaScriptExecutor.executeScript("return 'getAllAngularTestabilities' in window && 0 === window.getAllAngularTestabilities().map(a => a.getPendingRequestCount() + a.getPendingTasks().length).reduce((a, b) => a + b, 0)").toString();
            return Boolean.valueOf(javaScriptExecutionResult);
        };
    }
}
```

```java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.events.AbstractWebDriverEventListener;
import org.openqa.selenium.support.events.WebDriverEventListener;
import org.openqa.selenium.support.ui.WebDriverWait;
import ua.rabota.frankenstein.conditions.AngularConditions;

import static com.codeborne.selenide.WebDriverRunner.getWebDriver;

public class AngularEventListener extends AbstractWebDriverEventListener implements WebDriverEventListener {
    @Override
    public void afterNavigateTo(String url, WebDriver driver) {
        waitForAngular();
    }

    @Override
    public void afterNavigateBack(WebDriver driver) {
        waitForAngular();
    }

    @Override
    public void afterNavigateForward(WebDriver driver) {
        waitForAngular();
    }

    @Override
    public void afterNavigateRefresh(WebDriver driver) {
        waitForAngular();
    }

    @Override
    public void beforeFindBy(By by, WebElement element, WebDriver driver) {
        waitForAngular();
    }

    @Override
    public void beforeClickOn(WebElement element, WebDriver driver) {
        waitForAngular();
    }

    private void waitForAngular() {
        WebDriverWait wait = new WebDriverWait(getWebDriver(), 10, 100);
        wait.until(AngularConditions.angularHasFinishedProcessing());
    }
}
```

and using it in tests

```java
class MyTest {

    @BeforeAll
    public static void setUpAll() {
        addListener(new AngularEventListener());
    }
}
```

**Comment by rkeeves** [comment]
Hi!

**TLDR: Beware of the new Selenium listeners! Your throwables will be caught, logged then discarded by selenium's `EventFiringDecorator`.**

### Problem

You create some js magic based `FluentWait` webdriverlistener.

The listener gets called, and it retries evaluating the js until it is true, or timeout is reached.

If it reaches timeout, it throws. (it's pointless to continue the test at this point because Angular took too much time).

By throwing you expect the **test to fail immediately**... but what happens is that **the test will happily continue.**

Basically: If your listener throws, it'll be simply logged and **selenium** (and **selenide**) will continue.

This might be good for you, or might be bad for you depending on your circumstances.

I just wanted to point out this behavior.

### Background (I might be wrong)

Selenium changed the listener interface and also how the event firing mechanism works a bit.

In Selenide (via `WebDriverRunner::addListener`) you basically hook up your listener to an `EventFiringDecorator`.

`EventFiringDecorator` catches any `Throwable` that you might've thrown. Logs it, then **continues the execution**.

The [Javadoc of EventFiringDecorator](https://www.selenium.dev/selenium/docs/api/java/org/openqa/selenium/support/events/EventFiringDecorator.html) says:

```
Listeners can't affect driver behavior too much. They can't throw any exceptions (they can, but the decorator suppresses these exceptions), can't prevent execution of the decorated methods, can't modify parameters and results of the methods.

Decorators that modify the behaviour of the underlying drivers should be implemented by extending WebDriverDecorator, not by creating sophisticated listeners.
```

I don't know how to properly solve this issue without writing a `WebDriverDecorator` and decorating a driver instance with it, then setting this decorator as driver for Selenide manually in your beforeall.


Btw, selenide is insanely good, and this issue thread gave me a lot of ideas, so thanks for both!

Bye

**Comment by asolntsev** [comment]
@rkeeves Sorry, but I still don't understand why all this complexity is needed.
Selenide is designed to that you don't need to wait for any Angular specific details. 
You just write "$.shouldBe(visible)` - and Selenide waits until the element gets visible, no matter what framework it's rendered by. 

If you have some flaky tests with angular application, it's probably better to investigate the real reason of flakiness?

**Comment by rkeeves** [comment]
@asolntsev  Thanks again for Selenide! It's awesome!

I think I wasn't clear with the wording of my comment:

The point of my comment was not to judge whether this approach is viable or not.

I just wanted to make sure that if anyone stumbles upon this issue they'll know that:
With the new Selenium listener anything you throw will be caught and won't end the test.
*(It's not the listener, but the decorator impl which catches.)*

Is this a problem for Selenium/Selenide? No. Absolutely.

I just thought that this kind of change in behavior deserves a comment at least.

P.S.: I didn't use this approach for Angular, so I can't comment on it.
As you see my point has nothing to do with Angular specifically.
It's about changes in how the listeners behave generally.

**Comment by asolntsev** [comment]
@rkeeves Yes, yes, I understand. :) Knowledge exchange is great, thank you for sharing the information about Selenium listeners. 
I didn't want to judge or criticize. Rather I wanted to remind other readers that the life is much simplier, and they probably don't need to bother with details at all.
